xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/doc/XWikiDocument.java

```diff
@@ -3446,11 +3457,150 @@ public void readObjectsFromForm(EditForm eform, XWikiContext context) throws XWi
         }
     }
 
+    /**
+     * Generate a map from the request parameters of the form '<spacename>.<classname>_<number>_<propertyname>' Keys of
+     * this map will be the reference '<spacename>.<classname>' to the Class (for example, 'XWiki.XWikiRights'), the
+     * content is a list where each element describe property for the object <number>. Element of the list is a map
+     * where key is <propertyname> and content is the array of corresponding values. Example with a list of HTTP
+     * parameters:
+     * <ul>
+     * <li>XWiki.XWikiRights_0_users=XWiki.Admin</li>
+     * <li>XWiki.XWikiRights_0_users=XWiki.Me</li>
+     * <li>XWiki.XWikiRights_0_groups=XWiki.XWikiAllGroup</li>
+     * <li>XWiki.XWikiRights_1_user=XWiki.Admin</li>
+     * <li>XWiki.XWikiUsers_1_name=Spirou</li>
+     * </ul>
+     * will result in the following map <code><pre>
+     * {
+     *   "XWiki.XWikiRights": {
+     *     "0": {
+     *       "users": ["XWiki.Admin", "XWiki.Me"],
+     *       "groups": ["XWiki.XWikiAllGroup"]
+     *     },
+     *     "1": {
+     *       "users": ["XWiki.Admin"]
+     *     }
+     *   ],
+     *   "XWiki.XWikiUsers": 
+     *     "1": {
+     *       "name": ["Spirou"]
+     *     }
+     *   ]
+     * }
+     * </pre></code>
+     * 
+     * @param request The input HTTP request that provides the parameters
+     * @return The map containing ordered data
+     */
+    private Map<String, SortedMap<Integer, Map<String, String[]>>> parseRequestUpdateOrCreate(XWikiRequest request,
+        XWikiContext context)
+    {
+        Map<String, SortedMap<Integer, Map<String, String[]>>> result = new HashMap<>();
+        @SuppressWarnings("unchecked")
+        Map<String, String[]> allParameters = request.getParameterMap();
+        for (Entry<String, String[]> parameter : allParameters.entrySet()) {
+            Matcher matcher = XPROPERTY_URL_REFERENCE.matcher(parameter.getKey());
+            if (matcher.matches() == false) {
+                continue;
+            }
+            String className;
+            String classNumberAsString;
+            Integer classNumber;
+            String classPropertyName;
+            className = matcher.group(1);
+            classNumberAsString = matcher.group(2);
+            classPropertyName = matcher.group(3);
```

Comment:
No need to separate declaration from initialization for these variables, i.e. just use:
`String className = matcher.group(1);`

...except for `classNumber`, of course, which is initialized in the try block below.